mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessageVaultHook.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.james.mailbox.store.SessionProvider;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.GroupedFlux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class DeletedMessageVaultHook implements PreDeletionHook {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeletedMessageVaultHook.class);
+
+    private final MailboxSession session;
+    private final DeletedMessageVault deletedMessageVault;
+    private final DeletedMessageConverter deletedMessageConverter;
+    private final MailboxSessionMapperFactory mapperFactory;
+
+    DeletedMessageVaultHook(SessionProvider sessionProvider,
+                            DeletedMessageVault deletedMessageVault,
+                            DeletedMessageConverter deletedMessageConverter,
+                            MailboxSessionMapperFactory mapperFactory) {
+        this.session = sessionProvider.createSystemSession(getClass().getName());
+        this.deletedMessageVault = deletedMessageVault;
+        this.deletedMessageConverter = deletedMessageConverter;
+        this.mapperFactory = mapperFactory;
+    }
+
+    @Override
+    public Publisher<Void> notifyDelete(DeleteOperation deleteOperation) {
+        Preconditions.checkNotNull(deleteOperation);
+
+        return groupMetadataByOwnerAndMessageId(deleteOperation)
+            .flatMap(Throwing.<DeletedMessageMetadata, Mono<Void>>function(entry -> moveIntoTheVault(entry)).sneakyThrow())
+            .then();
+    }
+
+    private Mono<Void> moveIntoTheVault(DeletedMessageMetadata deletedMessageMetadata) throws MailboxException, IOException {
+        Optional<MailboxMessage> mailboxMessage = mapperFactory.getMessageIdMapper(session).find(ImmutableList.of(deletedMessageMetadata.getMessageId()), MessageMapper.FetchType.Full).stream().findFirst();
+
+        if (mailboxMessage.isPresent()) {
+            DeletedMessage deletedMessage = deletedMessageConverter.convert(deletedMessageMetadata, mailboxMessage.get());
+            return Mono.from(deletedMessageVault.append(deletedMessage.getOwner(), deletedMessage, mailboxMessage.get().getFullContent()));
+        }
+        LOGGER.warn("Can not find mailbox message with msgId {}", deletedMessageMetadata.getMessageId());
+
+        return Mono.empty();
+    }
+
+    private Flux<DeletedMessageMetadata> groupMetadataByOwnerAndMessageId(DeleteOperation deleteOperation) {
+        return Flux.fromIterable(deleteOperation.getDeletionMetadataList())
+            .publishOn(Schedulers.elastic())
+            .map(Throwing.function(this::toMetadataWithOwner))
```

Comment:
what about grouping by mailboxId at this point to avoid calling `findMailboxById` again and again ?

Reply:
I think grouping by mailboxId have no meaning at this case. We want to combine the message is reference in mailboxes before convert to a DeletedMessage.

We only get the owner through `findMailboxById`.

Reply:
Let's:

 - Do a first groupBy by mailboxId
 - Retrieve for each mailboxId the user and flatten the sream. At that point we have a `Stream<Tuple3<MessageId, MailboxId, User>>`
 - Then we can groupBy `MessageId + Owner` to get the final representation

This would be equivalent but significantly more performant (think what deleting a mailbox with thousands of messages will be like!)